### PR TITLE
change(web): engine/keyboard-cache => engine/package-cache, modularizes model cache therein 🧩

### PR DESCRIPTION
**Pull request by jahorton**
Right, the embedded platforms _do_ utilize our predictive-text models..

Given how small the resulting module is and how it's similar in purpose to the keyboard-cache + its classes, the most sensible option is to put the new module in the same child project.  Which then further exacerbates the issue with that child project's name.  While `engine/package-cache` is a bit of a misnomer for KMW, the name fits very well thematically when compared to all our other platforms and its purpose should be immediately clear to anyone on the development team.

Note that the module does not, itself, load keyboards or models.  It _does_ store loaded keyboards and specifications for registered (loadable) models.  The actual resource-loading code is elsewhere for both resource types.

 @keymanapp-test-bot skip

### CODE REVIEW

File: web/src/engine/main/src/keymanEngine.ts [hunk 1]
```diff
@@ -190,6 +194,42 @@ export default class KeymanEngine<ContextManager extends ContextManagerBase, Har
...
+
+  // 17.0: new!  Only used by apps utilizing app/webview and one app/browser test page.
+  removeModel(modelId: string) {
+    this.modelCache.deregister(modelId);
+
+    // Is it the active model?
+    if(this.processor.activeModel && this.processor.activeModel.id == modelId) {
+      this.processor.languageProcessor.unloadModel();
+    }
+  }
```
Comment (mcdurdin) [inline web/src/engine/main/src/keymanEngine.ts:232]: We need good API documentation if these are being published, ideally a javadoc style comment for eacch endpoint as well as the corresponding docs in help.keyman.com

File: web/src/engine/main/src/keymanEngine.ts [hunk 2]
```diff
@@ -190,6 +194,42 @@ export default class KeymanEngine<ContextManager extends ContextManagerBase, Har
...
+    if(this.processor.activeModel != model) {
+      if(this.processor.activeModel) {
+        this.processor.languageProcessor.unloadModel();
+      }
+
+      if(model) {
+        this.processor.languageProcessor.loadModel(model);
+      }
+    }
+  }
```
Comment (mcdurdin) [inline web/src/engine/main/src/keymanEngine.ts:211]: Why is this change needed?

File: web/src/engine/package-cache/src/modelCache.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      if(!code) {
+        console.warn("Null / undefined language codes are not permitted for registration.");
+        return;
+      }
+
+      mm.languageModelMap[code] = model;
+    });
+  }
+
+  deregister(modelId: string): ModelSpec {
```
Comment (mcdurdin) [inline web/src/engine/package-cache/src/modelCache.ts:39]: Probably should be `unregister` because we use that terminology a bit with keyboards...

**Review by mcdurdin — APPROVED** [review]

File: web/src/engine/main/src/keymanEngine.ts (hunk 2 above)
Comment (jahorton) [reply web/src/engine/main/src/keymanEngine.ts:211]: The old "ModelManager" class from namespaced-code was both a "model cache" and "model activator".  These two roles are split by this PR.  The `ModelCache` class maintains the former, while `KeymanEngine` itself will handle the latter, with this method as the primary method to do exactly that.

Compare this with the old `ModelManager.onKeyboardChange` method.  This method is pretty much just a renamed version of that, but based in a different location (`KeymanEngine`) than before.

File: web/src/engine/package-cache/src/modelCache.ts (hunk 3 above)
Comment (jahorton) [reply web/src/engine/package-cache/src/modelCache.ts:39]: Can do.

On a whim, I decided to check on the difference between "unregister" and "deregister":

https://english.stackexchange.com/questions/25931/unregister-vs-deregister

... yeah, let's leave that can of worms unopened.

File: web/src/engine/main/src/keymanEngine.ts (hunk 1 above)
Comment (jahorton) [reply web/src/engine/main/src/keymanEngine.ts:232]: I don't think we plan to publish these outright as of yet; after all, they're only really meant for use by our mobile apps.

That said, it is a decent prototype toward something we _could_ publish, so the least I can do is provide that Javadoc-style comment now - that's definitely a fair request.
